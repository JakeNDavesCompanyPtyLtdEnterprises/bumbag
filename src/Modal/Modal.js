// @flow
import React, { Fragment, type Node } from 'react';
import Portal from 'reakit/Portal';

import Backdrop from '../Backdrop/Backdrop.tsx';
import TrapFocus from '../_utils/TrapFocus';
import ModalContainer from './ModalContainer';
import _Modal, { ModalHide, ModalShow } from './styled';

type Props = {
  children: Node | Function,
  className?: string,
  /** Delay of the animation if one is specified. */
  delay?: string,
  /** Duration of the animation if one is specified. */
  duration?: string,
  /** Will the modal have an expand animation when it is toggled on/off? */
  expand?: boolean | string,
  /** Will the modal have a fade animation when it is toggled on/off? */
  fade?: boolean,
  hide: Function,
  /** Should the modal be hidden when 'esc' is clicked? */
  hideOnEsc?: boolean,
  /** Should the modal be hidden when the outside has been clicked? */
  hideOnClickOutside?: boolean,
  /** Whether or not to show the modal component */
  isVisible?: boolean,
  showActionButtons?: boolean,
  showCloseButton?: boolean,
  /** Will the modal have a slide animation when it is toggled on/off? */
  slide?: boolean | string,
  /** Timing of the animation */
  timing?: string,
  type?: ?'alert'
};

const Modal = ({
  children,
  delay,
  duration,
  fade,
  hide,
  hideOnEsc,
  hideOnClickOutside,
  isVisible,
  showActionButtons,
  showCloseButton,
  type,
  ...props
}: Props) => (
  <Portal>
    <TrapFocus isActive={isVisible} usesPortal>
      {({ fallbackFocusRef, initialFocusRef }) => (
        <Fragment>
          <Backdrop
            delay={delay}
            duration={duration}
            expand={undefined}
            hide={hideOnClickOutside && type !== 'alert' ? hide : undefined}
            fade={fade}
            hideOnEsc={type === 'alert' ? false : hideOnEsc}
            isVisible={isVisible}
            slide={undefined}
            use={hideOnClickOutside && type !== 'alert' ? ModalHide : undefined}
          />
          <_Modal
            aria-modal="true"
            delay={delay}
            duration={duration}
            fade={fade}
            hideOnEsc={type === 'alert' ? false : hideOnEsc}
            hideOnClickOutside={type === 'alert' ? false : hideOnClickOutside}
            isVisible={isVisible}
            {...props}
          >
            {typeof children === 'function' ? children({ fallbackFocusRef, initialFocusRef }) : children}
          </_Modal>
        </Fragment>
      )}
    </TrapFocus>
  </Portal>
);

Modal.defaultProps = {
  className: undefined,
  delay: undefined,
  duration: '250ms',
  expand: undefined,
  fade: false,
  hideOnEsc: true,
  hideOnClickOutside: true,
  isVisible: false,
  showActionButtons: false,
  showCloseButton: false,
  slide: false,
  timing: 'ease-in-out',
  type: undefined
};

Modal.Container = ModalContainer;
Modal.Hide = ModalHide;
Modal.Show = ModalShow;

export default Modal;
