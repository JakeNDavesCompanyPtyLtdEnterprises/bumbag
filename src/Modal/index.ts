export { default } from './Modal';
export { default as Modal } from './Modal';
export { default as ModalContainer } from './ModalContainer';
export { default as ModalHide } from './ModalHide';
export { default as ModalShow } from './ModalShow';
