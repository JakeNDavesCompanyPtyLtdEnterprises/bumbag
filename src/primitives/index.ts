// @ts-ignore
export { default as Box } from 'reakit/Box';
// @ts-ignore
export { default as Block } from 'reakit/Block';
// @ts-ignore
export { default as Inline } from 'reakit/Inline';
// @ts-ignore
export { default as InlineBlock } from 'reakit/InlineBlock';
// @ts-ignore
export { default as Flex } from 'reakit/Flex';
// @ts-ignore
export { default as InlineFlex } from 'reakit/InlineFlex';
// @ts-ignore
export { default as Grid } from 'reakit/Grid';
