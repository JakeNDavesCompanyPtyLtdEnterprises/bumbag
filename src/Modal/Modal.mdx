---
name: Modal
route: /utils/modal
menu: Utilities
---

import { Playground, PropsTable } from 'docz';
import { Box } from '../primitives/index';
import Hidden from 'reakit/Hidden';
import { Modal } from './Modal';
import Button, { ActionButtons } from '../Button';
import Pane from '../Pane';

# Modal

## Import

`import { Modal } from 'fannypack'`

## Basic Usage

The `<Modal>` component uses `<Overlay>` internally, but it is rendered inside a portal with it's focus trapped (keyboard navigation) for accessibility and also includes a backdrop.

It is recommended to try use the [`<DialogModal>` component](/components/dialogmodal) before you use `<Modal>`. `<Modal>` provides a way to create a more flexible modal if you need. If `<DialogModal>` doesn't suit your needs, then feel free to use this.

It is recommended to use the modal helper components (`<Modal.Show>`, `<Modal.Hide>`) when building modals to leverage accessibility features they provide like so:

<Playground>
  <Modal.Container>
    {modal => (
      <Box>
        <Button use={Modal.Show} {...modal}>
          Click me
        </Button>
        <Modal {...modal}>
          {({ fallbackFocusRef, initialFocusRef }) => (
            <Pane elementRef={fallbackFocusRef} padding="small" isFullWidth>
              Hello world!
              <Button use={Modal.Hide} elementRef={initialFocusRef} {...modal}>
                Close
              </Button>
            </Pane>
          )}
        </Modal>
      </Box>
    )}
  </Modal.Container>
</Playground>

## `<Modal>` Props

<PropsTable of={Modal} />

## Theming

### Schema

```jsx
{
  base: string | Object,
  Hide: {
    base: string | Object
  },
  Show: {
    base: string | Object
  }
}
```
