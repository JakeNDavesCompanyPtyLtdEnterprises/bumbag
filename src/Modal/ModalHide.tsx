import * as React from 'react';
import {
  LocalOverlayHideProps,
  OverlayHideProps,
  overlayHideDefaultProps,
  overlayHidePropTypes
} from '../Overlay/OverlayHide';

import { ModalHide as _ModalHide } from './styled';

export interface LocalModalHideProps extends LocalOverlayHideProps {}
export type ModalHideProps = OverlayHideProps;

const ModalHide: React.FunctionComponent<LocalModalHideProps> = ({ children, ...props }) => (
  <_ModalHide {...props}>{children}</_ModalHide>
);

ModalHide.propTypes = {
  ...overlayHidePropTypes
};
ModalHide.defaultProps = {
  ...overlayHideDefaultProps
};

// @ts-ignore
const C: React.FunctionComponent<ModalHideProps> = ModalHide;
export default C;
