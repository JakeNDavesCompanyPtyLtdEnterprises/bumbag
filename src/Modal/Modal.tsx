import * as React from 'react';
// @ts-ignore
import PropTypes from 'prop-types';
// @ts-ignore
import Portal from 'reakit/Portal';

import { isFunction } from '../_utils/assert';
import TrapFocus from '../_utils/TrapFocus';
import Backdrop from '../Backdrop';
import { OverlayProps } from '../Overlay/Overlay';
import ModalContainer, { ModalContainerProps } from './ModalContainer';
import ModalHide, { ModalHideProps } from './ModalHide';
import ModalShow, { ModalShowProps } from './ModalShow';
import _Modal from './styled';
import {
  AnimateProps,
  animatePropTypes,
  animateDefaultProps,
  RestrictHideProps,
  restrictDefaultProps,
  restrictHidePropTypes
} from '../types';

export interface Props {
  children:
    | React.ReactNode
    | ((
        {
          fallbackFocusRef,
          initialFocusRef
        }: { fallbackFocusRef: React.RefObject<HTMLElement>; initialFocusRef: React.RefObject<HTMLElement> }
      ) => React.ReactNode);
  className?: string;
  /** Whether or not to show the modal component */
  isVisible?: boolean;
  hide?(): void;
  showActionButtons?: boolean;
  showCloseButton?: boolean;
  type?: 'alert' | void;
}
export type LocalModalProps = Props & RestrictHideProps & AnimateProps;
export type ModalProps = OverlayProps & LocalModalProps;
export interface ModalComponents {
  Container: React.FunctionComponent<ModalContainerProps>;
  Hide: React.FunctionComponent<ModalHideProps>;
  Show: React.FunctionComponent<ModalShowProps>;
}

export const Modal: React.FunctionComponent<LocalModalProps> & ModalComponents = ({
  children,
  delay,
  duration,
  fade,
  hide,
  hideOnEsc,
  hideOnClickOutside,
  isVisible,
  showActionButtons,
  showCloseButton,
  type,
  ...props
}) => (
  <Portal>
    <TrapFocus isActive={isVisible} usesPortal>
      {({ fallbackFocusRef, initialFocusRef }) => (
        <React.Fragment>
          <Backdrop
            delay={delay}
            duration={duration}
            expand={undefined}
            hide={hideOnClickOutside && type !== 'alert' ? hide : undefined}
            fade={fade}
            hideOnEsc={type === 'alert' ? false : hideOnEsc}
            isVisible={isVisible}
            slide={undefined}
            use={hideOnClickOutside && type !== 'alert' ? ModalHide : undefined}
          />
          <_Modal
            aria-modal="true"
            delay={delay}
            duration={duration}
            fade={fade}
            hideOnEsc={type === 'alert' ? false : hideOnEsc}
            hideOnClickOutside={type === 'alert' ? false : hideOnClickOutside}
            isVisible={isVisible}
            {...props}
          >
            {isFunction(children) ? children({ fallbackFocusRef, initialFocusRef }) : children}
          </_Modal>
        </React.Fragment>
      )}
    </TrapFocus>
  </Portal>
);

Modal.propTypes = {
  children: PropTypes.func,
  className: PropTypes.string,
  isVisible: PropTypes.bool,
  hide: PropTypes.func,
  showActionButtons: PropTypes.bool,
  showCloseButton: PropTypes.bool,
  type: PropTypes.oneOf(['alert']),
  ...animatePropTypes,
  ...restrictHidePropTypes
};
Modal.defaultProps = {
  className: undefined,
  isVisible: false,
  showActionButtons: false,
  showCloseButton: false,
  type: undefined,
  ...animateDefaultProps,
  ...restrictDefaultProps,
  hideOnClickOutside: true,
  hideOnEsc: true
};

Modal.Container = ModalContainer;
Modal.Hide = ModalHide;
Modal.Show = ModalShow;

const C: React.FunctionComponent<ModalProps> & ModalComponents = Modal;
export default C;
