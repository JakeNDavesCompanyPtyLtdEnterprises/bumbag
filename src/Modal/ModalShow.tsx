import * as React from 'react';
import {
  LocalOverlayShowProps,
  OverlayShowProps,
  overlayShowDefaultProps,
  overlayShowPropTypes
} from '../Overlay/OverlayShow';

import { ModalShow as _ModalShow } from './styled';

export interface LocalModalShowProps extends LocalOverlayShowProps {}
export type ModalShowProps = OverlayShowProps;

const ModalShow: React.FunctionComponent<LocalModalShowProps> = ({ children, ...props }) => (
  <_ModalShow {...props}>{children}</_ModalShow>
);

ModalShow.propTypes = {
  ...overlayShowPropTypes
};
ModalShow.defaultProps = {
  ...overlayShowDefaultProps
};

export default ModalShow as React.FunctionComponent<ModalShowProps>;
