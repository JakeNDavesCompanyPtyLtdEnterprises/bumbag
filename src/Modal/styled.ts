import { theme } from 'styled-tools';
import Overlay from '../Overlay';
import styled from '../styled';
import { ModalHideProps } from './ModalHide';
import { ModalShowProps } from './ModalShow';
import { ModalProps } from './Modal';

export const ModalHide = styled(Overlay.Hide)<ModalHideProps>`
  ${theme('fannypack.Modal.Hide.base')};
`;

export const ModalShow = styled(Overlay.Show)<ModalShowProps>`
  ${theme('fannypack.Modal.Show.base')};
`;

export default styled(Overlay)<ModalProps>`
  padding: 0 1rem;
  max-width: 600px;
  width: 100%;

  ${theme('fannypack.Modal.base')};
`;
