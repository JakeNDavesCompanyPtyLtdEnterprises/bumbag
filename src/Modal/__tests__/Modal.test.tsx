import * as React from 'react';
import render from '../../_utils/tests/render';
import { Box } from '../../primitives';
import Modal from '../Modal';
jest.mock('reakit/Portal', () => 'portal');

it('renders correctly for a default modal', () => {
  const { container } = render(
    <Modal.Container defaultVisible>
      {modal => (
        <Modal {...modal}>
          {({ fallbackFocusRef }) => <Box elementRef={fallbackFocusRef}>This is the content</Box>}
        </Modal>
      )}
    </Modal.Container>
  );
  expect(container.firstChild).toMatchSnapshot();
});
