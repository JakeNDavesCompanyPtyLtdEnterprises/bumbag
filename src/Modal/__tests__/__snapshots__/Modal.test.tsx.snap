// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly for a default modal 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c0 {
  -webkit-transform: translate3d(0px,0px,0px);
  -ms-transform: translate3d(0px,0px,0px);
  transform: translate3d(0px,0px,0px);
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 998;
  -webkit-tap-highlight-color: rgba(0,0,0,0);
  -moz-tap-highlight-color: rgba(0,0,0,0);
  background-color: black;
  opacity: 0.4;
}

.c0[aria-hidden="true"] {
  pointer-events: none;
  display: none !important;
}

.c3 {
  -webkit-transform: translate3d(-50%,-50%,0px);
  -ms-transform: translate3d(-50%,-50%,0px);
  transform: translate3d(-50%,-50%,0px);
  position: fixed;
  z-index: 19900410;
  left: 50%;
  top: 50%;
}

.c3[aria-hidden="true"] {
  pointer-events: none;
  display: none !important;
}

.c2 {
  padding: 0 1rem;
  max-width: 600px;
  width: 100%;
}

<portal>
  <div>
    <button
      aria-hidden="false"
      class="c0 c1 "
      role="button"
      tabindex="-1"
    />
    <div
      aria-hidden="false"
      aria-modal="true"
      class="c2 c3 c1"
      role="dialog"
    >
      <div
        class="c1"
      >
        This is the content
      </div>
    </div>
  </div>
</portal>
`;
